opm/simulators/wells/WellInterface_impl.hpp

```diff
@@ -1507,13 +1507,41 @@ namespace Opm
             return std::vector<Scalar>(this->num_components_, 0.0);
 
         double d = computeConnectionDFactor(perf, intQuants, ws);
-        const PhaseUsage& pu = this->phaseUsage();
-        double Q = std::abs(ws.perf_data.phase_rates[perf*pu.num_phases + pu.phase_pos[Gas]]);
+        if (d < 1.0e-15)  {
+            return wi;
+        }
+
+        // Solve a quadratic equation for a connection rate satisfying the ipr and the flow-dependent skin,
+        // then use this rate to evaluate the actual skin.
         const auto& connection = this->well_ecl_.getConnections()[ws.perf_data.ecl_index[perf]];
-        double Kh = connection.Kh();
-        double scaling = 3.141592653589 * Kh * connection.wpimult();
+        const double Kh = connection.Kh();
+        const double scaling = 3.141592653589 * Kh * connection.wpimult();
         const unsigned gas_comp_idx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
-        wi[gas_comp_idx]  = 1.0/(1.0/(trans_mult * this->well_index_[perf]) + (Q/2 * d / scaling));
+
+        const double connection_pressure = ws.perf_data.pressure[perf];
+        const double cell_pressure = getValue(intQuants.fluidState().pressure(FluidSystem::gasPhaseIdx));
+        const double drawdown = cell_pressure - connection_pressure;
+        const double invB = getValue(intQuants.fluidState().invB(FluidSystem::gasPhaseIdx));
+        const double mob_g = getValue(intQuants.mobility(FluidSystem::gasPhaseIdx)) * invB;
+        const double a = d;
+        const double b = 2*scaling/wi[gas_comp_idx];
+        const double c = -1.0*b*trans_mult*this->well_index_[perf]*mob_g*drawdown;
+
```

Comment:
I suggest using
c = -2*scaling*mob_g*drawdown
to improve readability of the code.

Reply:
Thanks, good point, will update.